sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/GlobalPartitionEndpointManagerForCircuitBreaker.java

```diff
@@ -0,0 +1,747 @@
+        checkNotNull(request.requestContext, "requestContext cannot be null!");
+
+        PartitionKeyRange partitionKeyRange = request.requestContext.resolvedPartitionKeyRange;
+
+        if (partitionKeyRange == null) {
+            return;
+        }
+
+        String resourceId = request.getResourceId();
+        checkNotNull(resourceId, "resourceId cannot be null!");
+
+        logger.info("Handling exception : {}", resourceId);
+
+        PartitionKeyRangeWrapper partitionKeyRangeWrapper = new PartitionKeyRangeWrapper(partitionKeyRange, resourceId);
+
+        AtomicBoolean isFailoverPossible = new AtomicBoolean(true);
+        AtomicBoolean isFailureThresholdBreached = new AtomicBoolean(false);
+
+        this.partitionKeyRangeToLocationSpecificUnavailabilityInfo.compute(partitionKeyRangeWrapper, (partitionKeyRangeWrapperAsKey, partitionLevelLocationUnavailabilityInfoAsVal) -> {
+
+            if (partitionLevelLocationUnavailabilityInfoAsVal == null) {
+                partitionLevelLocationUnavailabilityInfoAsVal = new PartitionLevelLocationUnavailabilityInfo();
+            }
+
+            isFailureThresholdBreached.set(partitionLevelLocationUnavailabilityInfoAsVal.handleException(partitionKeyRangeWrapperAsKey, failedLocation, request.isReadOnlyRequest()));
+
+            if (isFailureThresholdBreached.get()) {
+
+                UnmodifiableList<URI> applicableEndpoints = request.isReadOnlyRequest() ?
+                    this.globalEndpointManager.getApplicableReadEndpoints(request.requestContext.getExcludeRegions()) :
+                    this.globalEndpointManager.getApplicableWriteEndpoints(request.requestContext.getExcludeRegions());
+
+                isFailoverPossible.set(
+                    partitionLevelLocationUnavailabilityInfoAsVal.areLocationsAvailableForPartitionKeyRange(partitionKeyRangeWrapperAsKey, applicableEndpoints, request.isReadOnlyRequest()));
+            }
+
+            return partitionLevelLocationUnavailabilityInfoAsVal;
+        });
+
+        // set to true if and only if failure threshold exceeded for the region
+        // and if failover is possible
+        // a failover is only possible when there are available regions left to fail over to
+        if (isFailoverPossible.get()) {
+            return;
+        }
+
+        // no regions to fail over to
+        this.partitionKeyRangeToLocationSpecificUnavailabilityInfo.remove(partitionKeyRangeWrapper);
+    }
+
+    public void handleLocationSuccessForPartitionKeyRange(RxDocumentServiceRequest request) {
+
+        checkNotNull(request, "request cannot be null!");
+        checkNotNull(request.requestContext, "requestContext cannot be null!");
+
+        PartitionKeyRange partitionKeyRange = request.requestContext.resolvedPartitionKeyRange;
+
+        if (partitionKeyRange == null) {
+            return;
+        }
+
+        String resourceId = request.getResourceId();
+        logger.info("Handling success : {}", resourceId);
```

Comment:
why info - for exception above that seems acceptable - but for success debug?

Reply:
Fixed in next iteration.